src/BaseKrylov.fypp

```diff
@@ -121,84 +129,153 @@ contains
         endif
 
         return
-    end subroutine initialize_krylov_subspace_${t1[0]}$${k1}$
+    end subroutine initialize_krylov_subspace_${type[0]}$${kind}$
+
+    subroutine orthogonalize_vector_against_basis_${type[0]}$${kind}$(y, X, info, beta)
+      !! Orthonormalizes the `abstract_vector` `y` against a basis `X` of `abstract_vector`.
+      class(abstract_vector_${type[0]}$${kind}$), intent(inout) :: y
+      !! Input `abstract_vector` to orthogonalize
+      class(abstract_vector_${type[0]}$${kind}$), intent(in)    :: X(:)
+      !! Input `abstract_vector` basis to orthogonalize against
+      integer, intent(out) :: info
+      !! Information flag.
+      ${type}$,                         optional, intent(out)   :: beta(:)
+      !! Projection coefficients if requested
+
+      ! internals
+      ${type}$ :: proj_coefficients(size(X))
+
+      info = 0
+
+      ! check for zero vector
+      if (y%norm() < atol_${kind}$) info = 1
+
+      ! orthogonalize
+      call innerprod(proj_coefficients, X, y)
+      block
+         class(abstract_vector_${type[0]}$${kind}$), allocatable :: proj
+         call linear_combination(proj, X, proj_coefficients)
+         call y%sub(proj)
+      end block
+
+      if (present(beta)) then
+         ! check size
+         call assert_shape(beta, [ size(X) ], 'orthogonalize_vector_against_basis_${type[0]}$${kind}$', 'beta')
+         beta = proj_coefficients
+      end if
+      
+      return
+    end subroutine orthogonalize_vector_against_basis_${type[0]}$${kind}$
+
+    subroutine orthogonalize_basis_against_basis_${type[0]}$${kind}$(Y, X, info, beta)
+      !! Orthonormalizes the `abstract_vector` basis `Y` against a basis `X` of `abstract_vector`.
```

Comment:
Same comments as for the basis vs. vector implementation.

Reply:
- `double_gram_schmidt_step` implemented & deployed.
- optional input flat `if_chk_orthonormal` added to `orthogonalize_against_basis`. Default is `.true.` for safety. Set to `.false.` in all performance-critical core functions. Appropriate logging is implemented.
- I deployed `orthogonalize_against_basis` in the lanczos solvers. I see that you only do a single GS pass there, so I left it that way for now. I think one should do a DGS pass though, it shouldn't impact performance much. Let me know if I should change it.